src/hooks/UseGraph/Graph.ts

```diff
@@ -0,0 +1,1147 @@
+   * Executes logic for each vertex of the graph
+   *
+   * @type {ForEachVertex<TData, TDimensions>}
+   */
+  forEachVertex: ForEachVertex<TData, TDimensions>
+
+  /**
+   * Returns `true` if calling the `callback` on any vertex returns `true`
+   *
+   * @type {SomeVertex<TData, TDimensions>}
+   */
+  someVertex: SomeVertex<TData, TDimensions>
+
+  /**
+   * Transforms all vertices in the graph
+   *
+   * @type {UpdateAllVertices<TData, TDimensions>}
+   */
+  updateAllVertices: UpdateAllVertices<TData, TDimensions>
+
+  /**
+   * Sets all vertices in the graph to a new value
+   *
+   * @type {SetAllVertices<TData>}
+   */
+  setAllVertices: SetAllVertices<TData>
+
+  /**
+   * Returns a graph where all `TData` vertices have been transformed to `TResult`s
+   *
+   * @type {MapAllVertices<TData, TDimensions>}
+   */
+  mapAllVertices: MapAllVertices<TData, TDimensions>
+}
+
+/**
+ * A data structure capable of containing a graph of any depth
+ *
+ * @export
+ * @class Graph
+ * @typedef {Graph}
+ * @template TData
+ * @template {number} [TDimensions=0]
+ * @implements {IGraph<TData, TDimensions>}
+ */
+export class Graph<TData, TDimensions extends number = 0> implements IGraph<TData, TDimensions> {
+  // #region Graph attributes
+  /**
+   * The internal data of the Graph
+   *
+   * @private
+   * @type {GraphData<TData, TDimensions>}
+   */
+  private data: GraphData<TData, TDimensions> | null = null
+
+  /**
+   * The number of dimensions in the graph
+   *
+   * @private
+   * @type {TDimensions}
+   */
+  private dimensions: TDimensions
+  // #endregion
+
+  // #region Graph constructors
+  /**
+   * Creates an instance of Graph with set dimensions
+   *
+   * @constructor
+   */
+  constructor(initial: { dimensions?: TDimensions })
+  /**
+   * Creates a Graph from existing graph data and dimensions
+   *
+   * @constructor
+   * @param {GraphData<TData, TDimensions>} data
+   */
+  constructor(initial: { dimensions: TDimensions; data: GraphData<TData, TDimensions> | null })
+  /**
+   * Copies a Graph into a new Graph
+   *
+   * @constructor
+   * @param {Graph<TData, TDimensions>} graph
+   */
+  constructor(initial: { graph: Graph<TData, TDimensions> })
+  /**
+   * Creates an instance of Graph
+   *
+   * @constructor
+   * @param {?GraphData<TData, TDimensions>} [data]
+   * @param {?Graph<TData, TDimensions>} [graph]
+   */
+  constructor({
+    graph,
+    dimensions,
+    data,
+  }: {
+    graph?: Graph<TData, TDimensions>
+    dimensions?: TDimensions
+    data?: GraphData<TData, TDimensions>
+  }) {
+    if (data && dimensions !== undefined) {
+      this.data = data
+      this.dimensions = dimensions
+    } else if (graph) {
+      this.data = graph.get()
+      this.dimensions = graph.getDimensions()
+    } else if (dimensions !== undefined) {
+      this.dimensions = dimensions
+    } else {
+      // @ts-expect-error TDimensions defaults to 0 if dimensions is not defined, so this is okay
+      this.dimensions = 0
```

Comment:
I think it might be better to cast here

```suggestion
      this.dimensions = 0 as TDimensions
```